nes-network/network/src/sender/channel.rs

```diff
@@ -0,0 +1,387 @@
+/*
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+use crate::channel::{Channel, Communication};
+use crate::protocol::*;
+use futures::SinkExt;
+use std::collections::{HashMap, VecDeque};
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio::select;
+use tokio::sync::oneshot;
+use tokio_stream::StreamExt;
+use tokio_util::sync::CancellationToken;
+use tracing::{Instrument, Span, debug, info, info_span, trace, warn};
+
+use super::control::*;
+pub type Result<T> = std::result::Result<T, Error>;
+pub type Error = Box<dyn std::error::Error + Send + Sync>;
+enum ChannelHandlerStatus {
+    ClosedByOtherSide,
+    ClosedBySoftware,
+    /// Like ClosedBySoftware, but the ChannelHandler failed to propagate to the other side.
+    ClosedBySoftwareButFailedToPropagate(Error),
+    Cancelled,
+}
+
+async fn create_channel<L: Communication>(
+    this_connection: ThisConnectionIdentifier,
+    target_identifier: ConnectionIdentifier,
+    channel_identifier: ChannelIdentifier,
+    communication: L,
+) -> Result<(
+    DataChannelSenderReader<L::Reader>,
+    DataChannelSenderWriter<L::Writer>,
+)> {
+    let stream = connect(target_identifier, communication).await?;
+    let (mut read, mut write) = identification_sender(stream);
+    write
+        .send(IdentificationRequest::IAmChannel(
+            this_connection,
+            channel_identifier,
+        ))
+        .await?;
+
+    let response = read
+        .next()
+        .await
+        .ok_or("Connection closed during identification")??;
+
+    let stream = Channel {
+        writer: write.into_inner().into_inner(),
+        reader: read.into_inner().into_inner(),
+    };
+
+    match response {
+        IdentificationResponse::Ok => Ok(data_channel_sender(stream)),
+        IdentificationResponse::NotOk => {
+            Err("Could not establish connection. Identification was denied".into())
```

Comment:
This is not used; under what circumstance would a peer deny the channel?

Reply:
Currently there is none, i removed the option